File: 백준/Silver/1932. 정수 삼각형/정수 삼각형.js
const fs = require('fs');
const input = fs.readFileSync(0, 'utf-8').toString().trim().split('\n');
const N = +input[0];
const triangle = input.slice(1).map(x => x.split(' ').map(Number));
const triangle_acc = Array.from(new Array(N), (_, i) => new Array(i + 1));
triangle_acc[0][0] = triangle[0][0];

for (let i = 1; i < N; i++) {
    for (let j = 0; j <= i; j++) {
        const left = triangle_acc[i - 1][j - 1] || 0;
        const right = triangle_acc[i - 1][j] || 0;
        
        triangle_acc[i][j] = Math.max(left, right) + triangle[i][j];
    }
}

console.log(Math.max(...triangle_acc[N - 1]));

File: 백준/Silver/1932. 정수 삼각형/정수 삼각형.py
import sys

N = int(input())
T = [[int(y) for y in x.split()] for x in sys.stdin.readlines()]
dp = []

for i in range(N):
    if i < 1:
        dp.append(T[0])
        continue
        
    line = []

    for j in range(i+1):
        left = dp[i-1][j-1] if j != 0 else 0
        right = dp[i-1][j] if j < i else 0
        line.append(max(left, right) + T[i][j])

    dp.append(line)

print(max(dp[N-1]))